File: app/buildz/tmp/kapt3/stubs/debug/com/bits/bee/bpmc/data/repository/RegRepositoryImpl.java
package com.bits.bee.bpmc.data.repository;

import java.lang.System;

/**
 * Created by aldi on 30/03/22.
 */
@kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u00000\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0010 \n\u0002\u0018\u0002\n\u0000\u0018\u00002\u00020\u0001B\u001f\b\u0007\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u0012\u0006\u0010\u0004\u001a\u00020\u0005\u0012\u0006\u0010\u0006\u001a\u00020\u0007\u00a2\u0006\u0002\u0010\bJ\u001a\u0010\t\u001a\u0014\u0012\u0010\u0012\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\r0\f0\u000b0\nH\u0016R\u000e\u0010\u0002\u001a\u00020\u0003X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0006\u001a\u00020\u0007X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082\u0004\u00a2\u0006\u0002\n\u0000\u00a8\u0006\u000e"}, d2 = {"Lcom/bits/bee/bpmc/data/repository/RegRepositoryImpl;", "Lcom/bits/bee/bpmc/domain/repository/RegRepository;", "apiUtils", "Lcom/bits/bee/bpmc/data/data_source/remote/ApiUtils;", "regDao", "Lcom/bits/bee/bpmc/data/data_source/local/dao/RegDao;", "ioDispatcher", "Lkotlinx/coroutines/CoroutineDispatcher;", "(Lcom/bits/bee/bpmc/data/data_source/remote/ApiUtils;Lcom/bits/bee/bpmc/data/data_source/local/dao/RegDao;Lkotlinx/coroutines/CoroutineDispatcher;)V", "getLatestRegList", "Lkotlinx/coroutines/flow/Flow;", "Lcom/bits/bee/bpmc/utils/Resource;", "", "Lcom/bits/bee/bpmc/domain/model/Reg;", "app_debug"})
public final class RegRepositoryImpl implements com.bits.bee.bpmc.domain.repository.RegRepository {
    private final com.bits.bee.bpmc.data.data_source.remote.ApiUtils apiUtils = null;
    private final com.bits.bee.bpmc.data.data_source.local.dao.RegDao regDao = null;
    private final kotlinx.coroutines.CoroutineDispatcher ioDispatcher = null;
    
    @javax.inject.Inject()
    public RegRepositoryImpl(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.data.data_source.remote.ApiUtils apiUtils, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.data.data_source.local.dao.RegDao regDao, @org.jetbrains.annotations.NotNull()
    kotlinx.coroutines.CoroutineDispatcher ioDispatcher) {
        super();
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<java.util.List<com.bits.bee.bpmc.domain.model.Reg>>> getLatestRegList() {
        return null;
    }
}

File: app/src/main/java/com/bits/bee/bpmc/data/repository/RegRepositoryImpl.kt
package com.bits.bee.bpmc.data.repository

import com.bits.bee.bpmc.data.data_source.local.dao.RegDao
import com.bits.bee.bpmc.data.data_source.remote.ApiUtils
import com.bits.bee.bpmc.data.data_source.remote.response.RegResponse
import com.bits.bee.bpmc.domain.mapper.RegDataMapper
import com.bits.bee.bpmc.domain.model.Reg
import com.bits.bee.bpmc.domain.repository.RegRepository
import com.bits.bee.bpmc.utils.ApiResponse
import com.bits.bee.bpmc.utils.NetworkDatabaseBoundResource
import com.bits.bee.bpmc.utils.Resource
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject

/**
 * Created by aldi on 30/03/22.
 */
class RegRepositoryImpl @Inject constructor(
    private val apiUtils: ApiUtils,
    private val regDao: RegDao,
    private val ioDispatcher: CoroutineDispatcher
) : RegRepository{

    override fun getLatestRegList(): Flow<Resource<List<Reg>>> {
        return object : NetworkDatabaseBoundResource<List<Reg>, RegResponse>(){
            override suspend fun loadFormDB(): List<Reg> {
                return regDao.getRegList().map { RegDataMapper.fromDbToDomain(it) }
            }

            override fun shouldFetch(data: List<Reg>?): Boolean {
                return true
            }

            override suspend fun createCall(): Flow<ApiResponse<RegResponse>> {
                return apiUtils.getRegApiService().getReg()
            }

            override suspend fun saveCallResult(data: RegResponse) {
                regDao.insertSingle(RegDataMapper.fromNetworkToDb(data.regModel))
            }
        }.getAsFlow()
    }

}